dataset/imagenet.py

```diff
@@ -0,0 +1,148 @@
+    ds = ds.map(lambda example: tf.io.parse_example(example, ImageNet.TFRECS_FORMAT))
+    ds = ds.map(lambda example: decode_example(example))
+    return ds
+
+
+  def _scale_and_center_crop(self, image, h, w, scale_size, final_size):
+    if w < h and w != scale_size:
+        w, h = tf.cast(scale_size, tf.int64), tf.cast(( (h / w) * scale_size), tf.int64)
+        im = tf.image.resize(image, (w,h))
+    elif h <= w and h != scale_size:
+        w, h = tf.cast(( (h / w) * scale_size), tf.int64), tf.cast(scale_size, tf.int64)
+        im = tf.image.resize(image, (w,h))
+    else:
+      im = tf.image.resize(image, (w,h))
+    x = (w - final_size) / 2 
+    y = (h - final_size) / 2
+    return im[y : (y + final_size), x : (x + final_size), :]
+
+
+  def _random_sized_crop(self, example, min_area = 0.08, max_iter = 10):
+    h, w = tf.cast(example['height'], tf.int64), tf.cast(example['width'], tf.int64)
+    area = h * w
+    return_example = False
+    image = example['image']
+    num_iter = tf.constant(0, dtype = tf.int32)
+    while num_iter <= max_iter:
+      num_iter  = tf.math.add(num_iter,1)
+      final_area = tf.random.uniform((), minval = min_area, maxval = 1,
+                                      dtype = tf.float32) * tf.cast(area, tf.float32)
+      aspect_ratio = tf.random.uniform((), minval = 3./4., maxval = 4./3.,
+                                      dtype = tf.float32)
+      w_cropped = tf.cast(tf.math.round(tf.math.sqrt(final_area * aspect_ratio)), tf.int64)
+      h_cropped = tf.cast(tf.math.round(tf.math.sqrt(final_area / aspect_ratio)), tf.int64)
+
+      if tf.random.uniform(()) < 0.5:
+        w_cropped, h_cropped = h_cropped, w_cropped
+      
+      if h_cropped <= h and w_cropped <= w:
+        return_example = True
+        if h_cropped == h: 
+          y = tf.constant(0, dtype = tf.int64)
+        else:
+          y = tf.random.uniform((),minval = 0, maxval = h - h_cropped, dtype = tf.int64)
+        if w_cropped == w:
+          x = tf.constant(0, dtype = tf.int64)
+        else:
+          x = tf.random.uniform((),minval = 0, maxval = w - w_cropped, dtype = tf.int64)
+      
+        image = image[y : (y + h_cropped), x : x + w_cropped, :]
+        image = tf.image.resize(image, (self.image_size, self.image_size))
+        break
+    if return_example:
+      return {
+          'image' : image,
+          'height' : self.image_size,
+          'width' : self.image_size,
+          'filename' : example['filename'],
+          'label' : example['label']
+        }
+
+    image = self._scale_and_center_crop(image, h, w, self.image_size, self.image_size)
+    return {
+          'image' : image,
+          'height' : self.image_size,
+          'width' : self.image_size,
+          'filename' : example['filename'],
+          'label' : example['label']
+        }
+  
+  @tf.function
+  def _pca_jitter(self, image):
+      image = tf.cast(image, tf.float32)/255.
```

Comment:
Instead of placing this scaling step here, I would recommend placing it in the function that will be used by all kinds of datasets i.e. train, validation, and test.

Reply:
Okay

Reply:
Or even better. Place it under the model itself. Keras now has the following layers:

* [Rescaling](https://www.tensorflow.org/api_docs/python/tf/keras/layers/experimental/preprocessing/Rescaling)
* [Normalization](https://www.tensorflow.org/api_docs/python/tf/keras/layers/experimental/preprocessing/Normalization)

During initializing the `Normalization` layer, we can set the `mean` and `var` arguments to the ImageNet stats. Note that it's variance as opposed to standard deviation (which is typical). So, we would need to square the std.

Reply:
Sure